webhooks/services/billing_processor.py

```diff
@@ -0,0 +1,301 @@
+import asyncio
+from typing import Any, Dict, List, NoReturn
+
+import orjson
+from fastapi import HTTPException
+
+from shared.constants import LAGO_API_KEY, LAGO_URL
+from shared.log_config import get_logger
+from shared.models.endorsement import TransactionTypes
+from shared.util.rich_async_client import RichAsyncClient
+from webhooks.models.billing_payloads import (
+    AttribBillingEvent,
+    CredDefBillingEvent,
+    CredentialBillingEvent,
+    LagoEvent,
+    ProofBillingEvent,
+    RevocationBillingEvent,
+    RevRegDefBillingEvent,
+    RevRegEntryBillingEvent,
+)
+from webhooks.services.webhooks_redis_serivce import WebhooksRedisService
+
+logger = get_logger(__name__)
+
+
+class BillingManager:
+    """
+    Class process billing events and send them to LAGO
+    """
+
+    def __init__(self, redis_service: WebhooksRedisService) -> None:
+        self.redis_service = redis_service
+
+        self._tasks: List[asyncio.Task] = []
+
+        self._pubsub = None
+
+        self._client = RichAsyncClient(
+            name="BillingManager", headers={"Authorization": f"Bearer {LAGO_API_KEY}"}
+        )
+
+    def start(self) -> None:
+        """
+        Start the billing manager
+        """
+        asyncio.create_task(
+            self._listen_for_billing_events(), name="Listen for new billing events"
+        )
+
+    async def stop(self) -> None:
+        """
+        Wait for tasks to complete and stop the billing manager
+        """
+        for task in self._tasks:
+            task.cancel()  # Request cancellation of the task
+            try:
+                await task  # Wait for the task to be cancelled
+            except asyncio.CancelledError:
+                pass
+        self._tasks.clear()  # Clear the list of tasks
+        logger.info("Billing manager stopped")
+
+        if self._pubsub:
+            self._pubsub.disconnect()
+            logger.info("Billing pubsub disconnected")
+
+    def are_task_running(self) -> bool:
+        """
+        Check if tasks are running
+        """
+        logger.debug("Checking if tasks are running")
+
+        if not self._pubsub:
+            logger.error("Pubsub is not running")
+
+        all_running = self._tasks and all(not task.done() for task in self._tasks)
+        logger.debug("All tasks running: {}", all_running)
+
+        if not all_running:
+            for task in self._tasks:
+                if task.done():
+                    logger.warning("Task `{}` is not running", task.get_name())
+        return self._pubsub and all_running
```

Comment:
I wonder if this will cause problems with the health check when billing processor isn't defined.

By default, `are_task_running` (typo btw -> tasks) -- by default, self._pubsub will be None, and all_running could also be none because there are no self._tasks.

The reason I'm pointing this out is because we don't have test coverage for the healthcheck endpoints (they need dependency overriding that I haven't figured out yet). So this could break the healthcheck without us knowing from tests. So just keep in mind that if Lago doesn't exist, then the BillingProcessor should not start up, and the "are_tasks_running" check which is called by the healthcheck should return TRUE when Lago is undefined and no tasks exist ...

Reply:
At the moment all I am doing is not passing events on to the service if there is no `apikey` so the service is running... right?